File: content/snippets/go/interview_lcci/08/03_magic_index.go
package main

import "fmt"

func findMagicIndex(nums []int) int {
	lo, hi := 0, len(nums)-1
	for lo <= hi {
		mid := lo + (hi-lo)/2
		if nums[mid] == mid {
			return mid
		}
		if nums[mid] < mid {
			lo = mid + 1
		} else {
			hi = mid - 1
		}
	}
	return -1
}

func main() {
	fmt.Println(findMagicIndex([]int{-1, 1, 3, 4, 6})) // 1
	fmt.Println(findMagicIndex([]int{0, 2, 3, 4, 5}))  // 0
}

File: content/snippets/typescript/interview_lcci/08/03_magic_index.ts
function findMagicIndex(nums: number[]): number {
    let lo = 0, hi = nums.length - 1;
    while (lo <= hi) {
        const mid = (lo + hi) >> 1;
        if (nums[mid] === mid) return mid;
        if (nums[mid] < mid) lo = mid + 1;
        else hi = mid - 1;
    }
    return -1;
}

console.log(findMagicIndex([-1, 1, 3, 4, 6])); // 1
console.log(findMagicIndex([0, 2, 3, 4, 5]));  // 0
